src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,407 @@
+		}
+	}
+
+	public void setStreamInitialSequence(InitialPositionInStream streamInitialSequence) {
+		setStreamInitialSequenceExtended(InitialPositionInStreamExtended.newInitialPosition(streamInitialSequence));
+	}
+
+	public void setStreamInitialSequenceExtended(InitialPositionInStreamExtended streamInitialSequence) {
+		Assert.notNull(streamInitialSequence, "'streamInitialSequence' must not be null");
+		this.streamInitialSequence = streamInitialSequence;
+	}
+
+	public void setIdleBetweenPolls(int idleBetweenPolls) {
+		this.idleBetweenPolls = Math.max(250, idleBetweenPolls);
+	}
+
+	public void setConsumerBackoff(int consumerBackoff) {
+		this.consumerBackoff = Math.max(1000, consumerBackoff);
+	}
+
+	/**
+	 * Sets the interval between 2 checkpoints.
+	 *
+	 * @param checkpointsInterval interval between 2 checkpoints (in milliseconds)
+	 */
+	public void setCheckpointsInterval(long checkpointsInterval) {
+		this.checkpointsInterval = checkpointsInterval;
+	}
+
+	@Override
+	public String toString() {
+		return "KclMessageDrivenChannelAdapter{consumerGroup='" + this.consumerGroup + '\'' + ", stream='" + this.stream + "'}";
+	}
+
+	private class RecordProcessorFactory implements ShardRecordProcessorFactory {
+		@Override
+		public ShardRecordProcessor shardRecordProcessor() {
+			return new RecordProcessor();
+		}
+	}
+
+	/**
+	 * Processes records and checkpoints progress.
+	 */
+	private class RecordProcessor implements ShardRecordProcessor {
+
+		private String shardId;
+
+		private long nextCheckpointTimeInMillis;
+
+		/** {@inheritDoc} */
+		@Override
+		public void initialize(InitializationInput initializationInput) {
+			this.shardId = initializationInput.shardId();
+			if (logger.isInfoEnabled()) {
+				logger.info("Initializing record processor for shard: " + this.shardId);
+			}
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void leaseLost(LeaseLostInput leaseLostInput) {
+			logger.info("Lost lease, so terminating.");
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void shardEnded(ShardEndedInput shardEndedInput) {
+			try {
+				logger.info("Reached shard end checkpointing.");
+				shardEndedInput.checkpointer().checkpoint();
+			}
+			catch (ShutdownException | InvalidStateException e) {
+				logger.error("Exception while checkpointing at shard end.  Giving up", e);
+			}
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void shutdownRequested(ShutdownRequestedInput shutdownRequestedInput) {
+			try {
+				logger.info("Scheduler is shutting down, checkpointing.");
+				shutdownRequestedInput.checkpointer().checkpoint();
+			}
+			catch (ShutdownException | InvalidStateException e) {
+				logger.error("Exception while checkpointing at requested shutdown.  Giving up", e);
+			}
+		}
+
+		/**
+		 * Process records performing retries as needed. Skip "poison pill" records.
+		 *
+		 * @param records Data records to be processed.
+		 */
+		private void processRecordsWithRetries(List<KinesisClientRecord> records) {
```

Comment:
The name and JavaDocs are not relevant any more since we have moved retry logic away from this class.

In addition I think we need to introduce a `batch` mode when we can send all the `List<KinesisClientRecord>` downstream as a single `payload`. Right, converted, but still as a list.

Might be a separate issue though...